thread_sync.c

```diff
@@ -1430,26 +1430,53 @@ struct rb_condvar {
  *
  *  ConditionVariable objects augment class Mutex. Using condition variables,
  *  it is possible to suspend while in the middle of a critical section until a
- *  resource becomes available.
+ *  condition is met, such as a resource becomes available.
+ *
+ *  Due to non-deterministic scheduling and spurious wake-ups, users of
+ *  condition variables should always use a separate boolean predicate (such as
+ *  reading from a boolean variable) to check if the condition is actually met
+ *  before starting to wait, and should wait in a loop, re-checking the
+ *  condition every time the ConditionVariable is waken up.  The idiomatic way
+ *  of using condition variables is calling the +wait+ method in an +until+
+ *  loop with the predicate as the loop condition.
  *
  *  Example:
  *
  *    mutex = Thread::Mutex.new
- *    resource = Thread::ConditionVariable.new
+ *
+ *    resource_available = false
+ *    condvar = Thread::ConditionVariable.new
  *
  *    a = Thread.new {
- *	 mutex.synchronize {
- *	   # Thread 'a' now needs the resource
- *	   resource.wait(mutex)
- *	   # 'a' can now have the resource
- *	 }
+ *      mutex.synchronize {
+ *        # Thread 'a' now needs the resource.
+ *        # Note:
+ *        # (1) Thread 'b' may be scheduled before thread 'a', and may run so
+ *        #     fast that it have already made the resource available before
+ *        #     thread 'a' start.  Therefore, we should check
+ *        #     'resource_available' before starting to wait.
+ *        # (2) The 'wait' method may spuriously wake up without signalling.
+ *        #     Therefore, we should recheck 'resource_available' after waking
+ *        #     up, and go back to wait if the condition is not actually met.
+ *        # (3) As soon as the resource is available, a third thread (not
+ *        #     necessarily waiting on 'condvar') may be scheduled first and
+ *        #     may have acquired the 'mutex' and consumed the resource
+ *        #     before thread 'a' wakes up.  This implies rechecking, too.
```

Comment:
I don't think the example demonstrates this. Is this note necessary here?

Reply:
It is an important issue to consider in real-world concurrent programming.  I can add another thread `a2` which behaves exactly like `a` and that will make the note (3) necessary.